geode-for-redis/src/main/java/org/apache/geode/redis/internal/eventing/EventDistributor.java

```diff
@@ -0,0 +1,125 @@
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.eventing;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.geode.annotations.VisibleForTesting;
+import org.apache.geode.cache.partition.PartitionListenerAdapter;
+import org.apache.geode.logging.internal.executors.LoggingThreadFactory;
+import org.apache.geode.redis.internal.commands.RedisCommandType;
+import org.apache.geode.redis.internal.data.RedisKey;
+
+public class EventDistributor extends PartitionListenerAdapter {
+
+  private final Map<RedisKey, Queue<EventListener>> listeners = new ConcurrentHashMap<>();
+
+  private final ScheduledThreadPoolExecutor timerExecutor =
+      new ScheduledThreadPoolExecutor(1,
+          new LoggingThreadFactory("GeodeForRedisEventTimer-", true));
+
+  private int keysRegistered = 0;
+
+  public EventDistributor() {
+    timerExecutor.setRemoveOnCancelPolicy(true);
+  }
+
+  public synchronized void registerListener(EventListener listener) {
+    for (RedisKey key : listener.keys()) {
+      listeners.computeIfAbsent(key, k -> new LinkedBlockingQueue<>()).add(listener);
+    }
+
+    if (listener.getTimeout() != 0) {
+      scheduleTimeout(listener, listener.getTimeout());
+    }
+
+    keysRegistered += listener.keys().size();
+  }
+
+  public void fireEvent(RedisCommandType command, RedisKey key) {
+    Queue<EventListener> listenerList = listeners.get(key);
+    if (listenerList == null) {
+      return;
+    }
+
+    for (EventListener listener : listenerList) {
+      if (listener.process(command, key) == EventResponse.REMOVE_AND_STOP) {
+        removeListener(listener);
+        break;
+      }
+    }
+  }
+
+  /**
+   * The total number of keys registered by all listeners (includes duplicates).
+   */
+  @VisibleForTesting
+  public int size() {
+    return keysRegistered;
+  }
+
+  @Override
+  public void afterBucketRemoved(int bucketId, Iterable<?> keys) {
+    Set<EventListener> resubmittingList = new HashSet<>();
+    for (Map.Entry<RedisKey, Queue<EventListener>> entry : listeners.entrySet()) {
+      if (entry.getKey().getBucketId() == bucketId) {
+        resubmittingList.addAll(entry.getValue());
+      }
+    }
+
+    resubmittingList.forEach(x -> {
+      x.resubmitCommand();
+      removeListener(x);
+    });
+  }
+
+  private synchronized void removeListener(EventListener listener) {
+    boolean listenerRemoved = false;
+    for (RedisKey key : listener.keys()) {
+      Queue<EventListener> listenerList = listeners.get(key);
+      if (listenerList == null) {
+        continue;
+      }
+      listenerRemoved |= listenerList.remove(listener);
+      if (listenerList.isEmpty()) {
```

Comment:
I think the only concurrency issue you have is these two lines (isEmpty and remove) and the add call in registerListener. If you got rid of these two lines then you would have no concurrency issues but you would leave empty queues in the map that may never be used again. But if you do this listeners.remove(key) call then you have an issue. The way I addressed this in the pubsub data structures was to create the queue non-empty in registerListener and if it every finds an empty queue it refuses to add to it and instead spins around and reads from the map again until it find a non-empty one or creates its own non-empty one. To make this thread safe you would create you own queue class whose add and remove inc and dec an atomic. The constructor of that class would always start with one element so the atomic would start at 1. If it ever goes to 0 that sticks.
This gets complicated so for now maybe you should keep some synchronization. You can make the sync issue more clear by narrowing your sync to be around just the putIfAbsent+add call and around this isEmpty+remove calls. Also you only need to do the sync(isEmpty+remove) calls if listenerList.remove returned true.

Reply:
I did keep the synchronization. Perhaps I'm mistaken, but I don't think these blocking commands are going to have any hot paths.